ARCHITECTURE.md

```diff
@@ -0,0 +1,339 @@
+This document describes the general design principles and modules of solid-client,
+to help contributors make sure their and others' work is coherent with what's
+there.
+
+For more information, be sure to use `git blame` liberally. We've generally kept
+fairly good Git hygiene, so reading the commit that introduced some code can be
+helpful in understanding how it relates to the rest of the codebase.
+
+# Design principles
+
+## Avoidance of side effects
+
+Generally, our API's tend to avoid side effects whenever possible. That means
+that we have a number of functions that only perform the side effects we cannot
+avoid (fetching data, storing data, ...), and all the rest just take some data
+as input and return modified copies of that data as output.
+
+The reason for this is that the front-end world has mostly converged to
+a model in which the relevant parts of the view are updated in response to
+changes in the data, and to verify whether that data has changed using
+referential equality (in other words, not comparing the content of the data,
+but whether it is a new instance) to avoid costly comparisons — e.g. Virtual
+DOM. Since Solid is all about the data, we avoid modifying the data in place to
+support this common use case.
+
```

Comment:
I think I'd sneak in an explicit reference to SolidDataset being a plain object (and not an RDFJS Dataset) here for posterity. This also relates to the 412 bug mentioned earlier, so maybe link the two sections somehow ?

Reply:
How about 0fc25cc9?